LIPs/LIP-service-registry.md

```diff
@@ -0,0 +1,88 @@
+
+Possible approaches for transcoders to advertise connection endpoints include DHT based discovery, a global registration HTTP endpoint and a smart contract registry. The downsides of DHT based discovery center around current unknowns around stability and engineering overhead. The downsides of a global registration HTTP endpoint center around the reliance on a centralized service. Consequently, a smart contract registry of service endpoints that transcoders can be use to publish an endpoint that broadcasters can send requests to appears to be a reasonable approach at this time.
+
+## Specification
+
+Create the contract ServiceRegistry with the following interface. A full implementation is provided in a later section of this proposal:
+
+```
+interface IServiceRegistry {
+    event ServiceURIUpdate(address indexed addr, string serviceURI);
+
+    function setServiceURI(string _serviceURI) external;
+    function getServiceURI(address _addr) public view returns (string);
+}
+```
+
+The contract uses a `Record` struct to store the service URI thus allowing for the addition of other state variables in the future if necessary.
+
+The contract uses a state variable `mapping (address => Record) private records` to manage the registry records.
+
+The function `setServiceURI(string _serviceURI)` does the following:
+- Store `_serviceURI` as the service URI endpoint for the caller that can be used to send requests to the caller off-chain
+- This function will only set the service URI endpoint for the caller and no other accounts
+- Successful execution of this function will emit the `ServiceURIUpdate()` event that will include the address of the caller and its new 
+service URI. The address is indexed thus allowing clients to filter for `ServiceURIUpdate()` events specifically for changes from a specific caller address
+
+The function `getServiceURI(address _addr)` does the following:
+- Return the service URI endpoint stored for the caller
+
+The contract is upgradeable using the delegate proxy pattern. Let `ServiceRegistryProxy` be the proxy contract and `ServiceRegistryTarget` be the target implementation contract.
+
+A block `upgradeBlock` and round `upgradeRound` (`upgradeBlock` must be a block in `upgradeRound`) are selected. At a block before `upgradeBlock` within `upgradeRound`, both `ServiceRegistryProxy` and `ServiceRegistryTarget` will be deployed with addresses `ServiceRegistryProxyAddress` and `ServiceRegistryTargetAddress` respectively. At block `upgradeBlock`, the owner of the Controller contract will call `Controller.setContract(contractId("ServiceRegistry"), ServiceRegistryProxyAddress)` and `Controller.setContract(contractId("ServiceRegistryTarget"), ServiceRegistryTargetAddress)`. After both of these Controller transactions are mined, the upgrade is complete and `ServiceRegistryProxy` will be registry contract used by transcoders to publish service URI endpoints.
```

Comment:
Probably because I'm unfamiliar with how the current contract upgrade procedure works, but:

Are `upgradeBlock` and `upgradeRound` actual values that are stored somewhere, or are they descriptive?

What happens if registry deployment and `Controller.setContract` happen to fall across different rounds?

Reply:
> Probably because I'm unfamiliar with how the current contract upgrade procedure works

I'll make a note to describe this process in more detail outside of this proposal!

`upgradeBlock` and `upgradeRound` are meant to be descriptive and not stored values. The idea was to attempt to offer a degree of transparency whenever this update happens so users have some understanding of when this feature will be included. Additionally, another idea is these values might be useful if clients want to embed logic that will switch to using this registry after some block or round.

It would be ok if the deployment of the registry and `Controller.setContract` fall across different rounds because clients will be using the Controller to look up the address of the registry so they wouldn't be trying to use the registry at least until the registry contract address can be looked up from the Controller. Furthermore, the system would likely be paused (no user transactions can be submitted) while this upgrade takes place and once the system is unpaused both the transactions would have been confirmed. However, I think ideally these two transactions occur within the same round which is reasonable since a round is currently a day.